File: apps/api/schemas/shadbala.py
"""
AstroOS — Shadbala API Schemas

Pydantic request/response models for the Shadbala endpoints. Mirrors
the request body shape used by divisional/dasha endpoints
(birth_datetime_utc, latitude, longitude, ayanamsa, house_system) since
ShadbalaEngine operates on the same D1Chart those engines build from
that data, and its cross-varga / sunrise-search sub-components
(Saptavargaja, Ojayugmarasyamsa, Tribhaga, Nathonnata, Dina-Hora Bala)
need exactly that same birth data again.

ShadbalaEngine deliberately does not expose a "total Shadbala" sum —
see services/shadbala_engine.py's module docstring: with some
sub-components still out of scope (Varsha/Masa lord), a sum would
misrepresent an incomplete result as complete. These schemas mirror
that: every component group is returned separately, never summed.
"""

from __future__ import annotations

from pydantic import BaseModel, Field

from apps.api.schemas.common import BirthDataInput


# ── Request ───────────────────────────────────────────────────────────────────


class ShadbalaRequest(BirthDataInput):
    """Request body for computing Shadbala components from birth data."""


# ── Response pieces ───────────────────────────────────────────────────────────


class BalaComponentResponse(BaseModel):
    """One Shadbala component's contribution for one planet."""

    component_id: str
    component_name: str
    rule_version: str
    planet: str
    value_shashtiamsas: float = Field(description="Value in Shashtiamsas (60ths of a Rupa).")
    trace: list[str] = Field(default_factory=list)


class Phase1ComponentsResponse(BaseModel):
    """Naisargika + Dig + Drik Bala."""

    naisargika_bala: list[BalaComponentResponse]
    dig_bala: list[BalaComponentResponse]
    drik_bala: list[BalaComponentResponse]


class Phase2ComponentsResponse(BaseModel):
    """Chesta + Paksha + Ayana + Yuddha Bala (Kala Bala sub-components)."""

    chesta_bala: list[BalaComponentResponse]
    paksha_bala: list[BalaComponentResponse]
    ayana_bala: list[BalaComponentResponse]
    yuddha_bala: list[BalaComponentResponse]


class SthanaBalaComponentsResponse(BaseModel):
    """Uchcha + Kendradi + Drekkana Bala (3 of Sthana Bala's 5 sub-components)."""

    uchcha_bala: list[BalaComponentResponse]
    kendradi_bala: list[BalaComponentResponse]
    drekkana_bala: list[BalaComponentResponse]


class SubBalaCheckResponse(BaseModel):
    """Pass/fail status of an individual sub-bala criteria."""

    bala_key: str
    bala_name: str
    obtained_virupas: float
    required_virupas: float
    passed: bool


class SaravaliPlanetSummaryResponse(BaseModel):
    """Complete Saravali Shadbala summary for a single planet."""

    planet: str
    planet_display_name: str

    # 6 Main Balas (Virupas)
    sthana_bala_virupas: float
    dig_bala_virupas: float
    kala_bala_virupas: float
    chesta_bala_virupas: float
    naisargika_bala_virupas: float
    drig_bala_virupas: float

    # Sthana Sub-components
    uchcha_bala_virupas: float
    saptavargaja_bala_virupas: float
    ojayugmarasyamsa_bala_virupas: float
    kendradi_bala_virupas: float
    drekkana_bala_virupas: float

    # Kala Sub-components
    nathonnata_bala_virupas: float
    paksha_bala_virupas: float
    tribhaga_bala_virupas: float
    dina_hora_bala_virupas: float
    ayana_bala_virupas: float
    yuddha_bala_virupas: float

    # Total Shadbala Pinda
    total_virupas: float
    total_rupas: float
    required_virupas: float
    required_rupas: float
    strength_ratio: float
    percentage: float
    is_strong: bool
    status_label: str
    rank: int

    # Ishta / Kashta
    ishta_bala_virupas: float
    kashta_bala_virupas: float

    # Individual Sub-Bala Checks
    sub_bala_checks: list[SubBalaCheckResponse]
    all_sub_balas_passed: bool


class SaravaliShadbalaReportResponse(BaseModel):
    """Complete aggregated Saravali Shadbala Report for all 7 classical grahas."""

    planets: list[SaravaliPlanetSummaryResponse]
    strongest_planet: str
    weakest_planet: str
    average_strength_ratio: float
    chart_strength_score: float


class AllShadbalaResponse(BaseModel):
    """
    Every implemented Shadbala component/sub-component, grouped exactly
    as ShadbalaEngine's compute_*() methods group them, plus full Saravali
    evaluation summary.
    """

    phase1: Phase1ComponentsResponse
    phase2: Phase2ComponentsResponse
    sthana_bala: SthanaBalaComponentsResponse
    saptavargaja_bala: list[BalaComponentResponse]
    ojayugmarasyamsa_bala: list[BalaComponentResponse]
    tribhaga_bala: list[BalaComponentResponse]
    nathonnata_bala: list[BalaComponentResponse]
    dina_hora_bala: list[BalaComponentResponse]
    ishta_bala: list[BalaComponentResponse] = Field(default_factory=list)
    kashta_bala: list[BalaComponentResponse] = Field(default_factory=list)
    implemented_components: list[str]
    not_yet_implemented_components: list[str]
    summary: SaravaliShadbalaReportResponse | None = None

File: apps/web/src/lib/shadbala.ts
/**
 * AstroOS — Saravali Shadbala API & Calculations
 *
 * Full 6-Fold Shadbala Suite matching Kalyana Varma's Saravali & BPHS Ch. 27:
 * 1. Sthana Bala (Uchcha, Saptavargaja, Ojayugmarasyamsa, Kendradi, Drekkana)
 * 2. Dig Bala (Directional Quadrants)
 * 3. Kala Bala (Nathonatha, Paksha, Tribhaga, Dina-Hora, Yudhdha, Ayana)
 * 4. Cheshta Bala (8 motional states & velocity)
 * 5. Naisargika Bala (Fixed luminosity scale)
 * 6. Drig Bala (Aspectual Sputa Drishti rectified)
 * Total Pinda (Virupas & Rupas), Minimum Requirements comparison, and Ishta/Kashta Bala.
 */

"use client";

import { useQuery } from "@tanstack/react-query";
import { api, tokenStore } from "./api";
import type { WorkflowAnalysisRequest } from "./types";

export interface BalaComponent {
  component_id: string;
  component_name: string;
  rule_version: string;
  planet: string;
  value_shashtiamsas: number;
  trace: string[];
}

export interface IshtaKashtaResponse {
  ishta_bala: BalaComponent[];
  kashta_bala: BalaComponent[];
}

export interface SubBalaCheck {
  bala_key: string;
  bala_name: string;
  obtained_virupas: number;
  required_virupas: number;
  passed: boolean;
}

export interface SaravaliPlanetSummary {
  planet: string;
  planet_display_name: string;

  // 6 Main Balas (Virupas)
  sthana_bala_virupas: number;
  dig_bala_virupas: number;
  kala_bala_virupas: number;
  chesta_bala_virupas: number;
  naisargika_bala_virupas: number;
  drig_bala_virupas: number;

  // Sthana Sub-components
  uchcha_bala_virupas: number;
  saptavargaja_bala_virupas: number;
  ojayugmarasyamsa_bala_virupas: number;
  kendradi_bala_virupas: number;
  drekkana_bala_virupas: number;

  // Kala Sub-components
  nathonnata_bala_virupas: number;
  paksha_bala_virupas: number;
  tribhaga_bala_virupas: number;
  dina_hora_bala_virupas: number;
  ayana_bala_virupas: number;
  yuddha_bala_virupas: number;

  // Total Shadbala Pinda
  total_virupas: number;
  total_rupas: number;
  required_virupas: number;
  required_rupas: number;
  strength_ratio: number;
  percentage: number;
  is_strong: boolean;
  status_label: "Strong" | "Moderate" | "Deficient" | string;
  rank: number;

  // Ishta / Kashta
  ishta_bala_virupas: number;
  kashta_bala_virupas: number;

  // Individual Sub-Bala Checks
  sub_bala_checks: SubBalaCheck[];
  all_sub_balas_passed: boolean;
}

export interface SaravaliShadbalaReport {
  planets: SaravaliPlanetSummary[];
  strongest_planet: string;
  weakest_planet: string;
  average_strength_ratio: number;
  chart_strength_score: number;
}

export interface Phase1ComponentsResponse {
  naisargika_bala: BalaComponent[];
  dig_bala: BalaComponent[];
  drik_bala: BalaComponent[];
}

export interface Phase2ComponentsResponse {
  chesta_bala: BalaComponent[];
  paksha_bala: BalaComponent[];
  ayana_bala: BalaComponent[];
  yuddha_bala: BalaComponent[];
}

export interface SthanaBalaComponentsResponse {
  uchcha_bala: BalaComponent[];
  kendradi_bala: BalaComponent[];
  drekkana_bala: BalaComponent[];
}

export interface AllShadbalaResponse {
  phase1: Phase1ComponentsResponse;
  phase2: Phase2ComponentsResponse;
  sthana_bala: SthanaBalaComponentsResponse;
  saptavargaja_bala: BalaComponent[];
  ojayugmarasyamsa_bala: BalaComponent[];
  tribhaga_bala: BalaComponent[];
  nathonnata_bala: BalaComponent[];
  dina_hora_bala: BalaComponent[];
  ishta_bala: BalaComponent[];
  kashta_bala: BalaComponent[];
  implemented_components: string[];
  not_yet_implemented_components: string[];
  summary?: SaravaliShadbalaReport | null;
}

// Classical Minimum Shadbala Constants (Saravali & BPHS Ch. 27)
export const SARAVALI_REQUIRED_VIRUPAS: Record<string, number> = {
  sun: 390.0,
  moon: 360.0,
  mars: 300.0,
  mercury: 420.0,
  jupiter: 390.0,
  venus: 330.0,
  saturn: 300.0,
};

export const SARAVALI_REQUIRED_RUPAS: Record<string, number> = {
  sun: 6.5,
  moon: 6.0,
  mars: 5.0,
  mercury: 7.0,
  jupiter: 6.5,
  venus: 5.5,
  saturn: 5.0,
};

export const SARAVALI_INDIVIDUAL_SUB_BALA_REQUIREMENTS: Record<
  string,
  { sthana: number; dig: number; kala: number; cheshta: number; ayana: number }
> = {
  sun: { sthana: 165, dig: 35, kala: 50, cheshta: 112, ayana: 30 },
  jupiter: { sthana: 165, dig: 35, kala: 50, cheshta: 112, ayana: 30 },
  mercury: { sthana: 165, dig: 35, kala: 50, cheshta: 112, ayana: 30 },
  moon: { sthana: 133, dig: 50, kala: 30, cheshta: 100, ayana: 40 },
  venus: { sthana: 133, dig: 50, kala: 30, cheshta: 100, ayana: 40 },
  mars: { sthana: 96, dig: 30, kala: 40, cheshta: 67, ayana: 20 },
  saturn: { sthana: 96, dig: 30, kala: 40, cheshta: 67, ayana: 20 },
};

export const NAISARGIKA_BALA_TABLE: Record<string, { virupas: number; rupas: string; rank: number }> = {
  sun: { virupas: 60.0, rupas: "7/7 (1.000)", rank: 1 },
  moon: { virupas: 51.43, rupas: "6/7 (0.857)", rank: 2 },
  venus: { virupas: 42.86, rupas: "5/7 (0.714)", rank: 3 },
  jupiter: { virupas: 34.29, rupas: "4/7 (0.571)", rank: 4 },
  mercury: { virupas: 25.71, rupas: "3/7 (0.429)", rank: 5 },
  mars: { virupas: 17.14, rupas: "2/7 (0.286)", rank: 6 },
  saturn: { virupas: 8.57, rupas: "1/7 (0.143)", rank: 7 },
};

export const shadbalaKeys = {
  ishtaKashta: (
    req: Pick<
      WorkflowAnalysisRequest,
      "birth_datetime_utc" | "latitude" | "longitude" | "ayanamsa" | "house_system"
    > | null,
  ) =>
    [
      "shadbala",
      "ishta-kashta",
      req?.birth_datetime_utc,
      req?.latitude,
      req?.longitude,
      req?.ayanamsa,
      req?.house_system,
    ] as const,
  all: (
    req: Pick<
      WorkflowAnalysisRequest,
      "birth_datetime_utc" | "latitude" | "longitude" | "ayanamsa" | "house_system"
    > | null,
  ) =>
    [
      "shadbala",
      "all",
      req?.birth_datetime_utc,
      req?.latitude,
      req?.longitude,
      req?.ayanamsa,
      req?.house_system,
    ] as const,
  summary: (
    req: Pick<
      WorkflowAnalysisRequest,
      "birth_datetime_utc" | "latitude" | "longitude" | "ayanamsa" | "house_system"
    > | null,
  ) =>
    [
      "shadbala",
      "summary",
      req?.birth_datetime_utc,
      req?.latitude,
      req?.longitude,
      req?.ayanamsa,
      req?.house_system,
    ] as const,
};

export function useIshtaKashtaBala(
  request: Pick<
    WorkflowAnalysisRequest,
    "birth_datetime_utc" | "latitude" | "longitude" | "ayanamsa" | "house_system"
  > | null,
) {
  return useQuery<IshtaKashtaResponse>({
    queryKey: shadbalaKeys.ishtaKashta(request),
    queryFn: () =>
      api.post<IshtaKashtaResponse>("/api/v1/shadbala/ishta-kashta", {
        birth_datetime_utc: request!.birth_datetime_utc,
        latitude: request!.latitude,
        longitude: request!.longitude,
        ayanamsa: request!.ayanamsa,
        house_system: request!.house_system,
      }),
    enabled: !!tokenStore.getAccess() && !!request,
    staleTime: Infinity,
  });
}

export function useShadbalaAll(
  request: Pick<
    WorkflowAnalysisRequest,
    "birth_datetime_utc" | "latitude" | "longitude" | "ayanamsa" | "house_system"
  > | null,
) {
  return useQuery<AllShadbalaResponse>({
    queryKey: shadbalaKeys.all(request),
    queryFn: () =>
      api.post<AllShadbalaResponse>("/api/v1/shadbala/all", {
        birth_datetime_utc: request!.birth_datetime_utc,
        latitude: request!.latitude,
        longitude: request!.longitude,
        ayanamsa: request!.ayanamsa,
        house_system: request!.house_system,
      }),
    enabled: !!tokenStore.getAccess() && !!request,
    staleTime: Infinity,
  });
}

export function useShadbalaSummary(
  request: Pick<
    WorkflowAnalysisRequest,
    "birth_datetime_utc" | "latitude" | "longitude" | "ayanamsa" | "house_system"
  > | null,
) {
  return useQuery<SaravaliShadbalaReport>({
    queryKey: shadbalaKeys.summary(request),
    queryFn: () =>
      api.post<SaravaliShadbalaReport>("/api/v1/shadbala/summary", {
        birth_datetime_utc: request!.birth_datetime_utc,
        latitude: request!.latitude,
        longitude: request!.longitude,
        ayanamsa: request!.ayanamsa,
        house_system: request!.house_system,
      }),
    enabled: !!tokenStore.getAccess() && !!request,
    staleTime: Infinity,
  });
}
